### ISSUE TITLE
Inconsistent accuracy evaluated with the provided pre-trained model 

### ISSUE THREAD

**Issue by qy-feng** [EricGuo5513/action-to-motion#4]
I could only get around 0.3 for class accuracy with the ground truth motion sequences using the provided pre-trained model files. Here are the logs:

==================== Replication 15 ====================
Time: 2021-02-28 19:28:50.781489
========== Evaluating Accuracy ==========
Calculating Accuracies...
---> [mocap] Accuracy: **0.3027**
Time: 2021-02-28 19:28:50.978719
Total number of frames 204079, videos 1088, action types 8
==================== Replication 16 ====================
Time: 2021-02-28 19:28:51.609661
========== Evaluating Accuracy ==========
Calculating Accuracies...
---> [mocap] Accuracy: **0.3053**
Time: 2021-02-28 19:28:51.813113
Total number of frames 204079, videos 1088, action types 8

**Comment by EricGuo5513** [comment]
Hi, are you using the dataloader in this repo? And are the accuracy for other datasets also abnormal?

**Comment by PACerv** [comment]
I think the issue might be how the accuracy is computed:
https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
The number of samples in the groundtruth dataset might not match `num_motions`.
I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
Any other ideas?

**Comment by EricGuo5513** [comment]
> I think the issue might be how the accuracy is computed:
> https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> 
> 
> The number of samples in the groundtruth dataset might not match `num_motions`.
> I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> Any other ideas?

Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.

**Comment by PACerv** [comment]
> > I think the issue might be how the accuracy is computed:
> > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > 
> > The number of samples in the groundtruth dataset might not match `num_motions`.
> > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > Any other ideas?
> 
> Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.

The motion_loaders are actually not included in the repository, so the details of the dataloader are unclear.

**Comment by qy-feng** [comment]
> I think the issue might be how the accuracy is computed:
> https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> 
> 
> The number of samples in the groundtruth dataset might not match `num_motions`.
> I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> Any other ideas?

The accuracy value seems normal after changing `num_motions`. I reimplemented the `motion_loader` by changing the data path to the path of generated files. However, I could only get 86% for HumanAct12 which is reported 92.3%.

**Comment by EricGuo5513** [comment]
> > > I think the issue might be how the accuracy is computed:
> > > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > > 
> > > The number of samples in the groundtruth dataset might not match `num_motions`.
> > > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > > Any other ideas?
> > 
> > 
> > Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.
> 
> The motion_loaders are actually not included in the repository, so the details of the dataloader are unclear.

Hi, we use this line for groundtruth motion_loader:
motion_loader = DataLoader(motion_dataset, batch_size=1, num_workers=1,
                               sampler=RandomSampler(motion_dataset, replacement=True, num_samples=num_motions))

Here, the **motion_dataset** is initialized as the same as it in the repo.

**Comment by EricGuo5513** [comment]
> > I think the issue might be how the accuracy is computed:
> > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > 
> > The number of samples in the groundtruth dataset might not match `num_motions`.
> > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > Any other ideas?
> 
> The accuracy value seems normal after changing `num_motions`. I reimplemented the `motion_loader` by changing the data path to the path of generated files. However, I could only get 86% for HumanAct12 which is reported 92.3%.

Hi, how do you manage the action categories during generation? If you are using the evaluation_motion_vae.py as in the tutorial, please use the following command instead:

python evaluate_motion_vae.py --name vanila_vae_tf --dataset_type humanact12  --motion_length 60 --coarse_grained --gpu_id 0 --do_random --num_samples 200 --name_ext R0

In our final evaluation, action category for each motion is randomly sampled instead of the deterministic one. You could change the value of --num_samples, the model would generate such number of motions in one batch.

**Comment by sjxAndy** [comment]
> > > > I think the issue might be how the accuracy is computed:
> > > > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > > > 
> > > > The number of samples in the groundtruth dataset might not match `num_motions`.
> > > > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > > > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > > > Any other ideas?
> > > 
> > > 
> > > Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.
> > 
> > 
> > The motion_loaders are actually not included in the repository, so the details of the dataloader are unclear.
> 
> Hi, we use this line for groundtruth motion_loader:
> motion_loader = DataLoader(motion_dataset, batch_size=1, num_workers=1,
> sampler=RandomSampler(motion_dataset, replacement=True, num_samples=num_motions))
> 
> Here, the **motion_dataset** is initialized as the same as it in the repo.

Hi, thank you for sharing your code for the paper! 

I had some trouble reproducing the quantitative results on NTU-RGBD dataset with lie. I found that if I set `num_samples` to `num_motions`, which is 13 for this dataset, the FID would be about 20, way higher than the result in the paper. I also tried setting `replacement` to False, and got FID around 0.76. 

Could you kindly provide a more detailed dataloader code for ground truth as well as the generated results? Thanks!

**Comment by EricGuo5513** [comment]
> > > > > I think the issue might be how the accuracy is computed:
> > > > > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > > > > 
> > > > > The number of samples in the groundtruth dataset might not match `num_motions`.
> > > > > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > > > > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > > > > Any other ideas?
> > > > 
> > > > 
> > > > Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.
> > > 
> > > 
> > > The motion_loaders are actually not included in the repository, so the details of the dataloader are unclear.
> > 
> > 
> > Hi, we use this line for groundtruth motion_loader:
> > motion_loader = DataLoader(motion_dataset, batch_size=1, num_workers=1,
> > sampler=RandomSampler(motion_dataset, replacement=True, num_samples=num_motions))
> > Here, the **motion_dataset** is initialized as the same as it in the repo.
> 
> Hi, thank you for sharing your code for the paper!
> 
> I had some trouble reproducing the quantitative results on NTU-RGBD dataset with lie. I found that if I set `num_samples` to `num_motions`, which is 13 for this dataset, the FID would be about 20, way higher than the result in the paper. I also tried setting `replacement` to False, and got FID around 0.76.
> 
> Could you kindly provide a more detailed dataloader code for ground truth as well as the generated results? Thanks!

Hi, the codes of Dataloaders & Datasets could be found in this repo [https://github.com/EricGuo5513/action2motion/tree/master/motion_loaders](https://github.com/EricGuo5513/action2motion/tree/master/motion_loaders) .Please note "Num Samples/Motions" means the number of motions we want to sample/generate. It should be 3000 for all datasets, and replacement is set to True in our experiments. "13" for NTU-RGBD only means the "Number of Action Categories".

**Comment by sjxAndy** [comment]
> > > > > > I think the issue might be how the accuracy is computed:
> > > > > > https://github.com/EricGuo5513/action-to-motion/blob/0f8d7c281af4c14dbf87756b37e7e1e3cf01ae4e/eval_scripts/final_evaluation.py#L21-L22
> > > > > > 
> > > > > > The number of samples in the groundtruth dataset might not match `num_motions`.
> > > > > > I've replace `num_motions` with `len(motion_loader.dataset)` to avoid this issue.
> > > > > > For cmu mocap my accuracy is still a few percent shy of the reported accuracy (~88%) though.
> > > > > > Any other ideas?
> > > > > 
> > > > > 
> > > > > Actually, even though num_motions is greater than len(groundtruth dataset), the dataloader would sample with replacement from the dataset. It should be fine to keep to original num_motions value.
> > > > 
> > > > 
> > > > The motion_loaders are actually not included in the repository, so the details of the dataloader are unclear.
> > > 
> > > 
> > > Hi, we use this line for groundtruth motion_loader:
> > > motion_loader = DataLoader(motion_dataset, batch_size=1, num_workers=1,
> > > sampler=RandomSampler(motion_dataset, replacement=True, num_samples=num_motions))
> > > Here, the **motion_dataset** is initialized as the same as it in the repo.
> > 
> > 
> > Hi, thank you for sharing your code for the paper!
> > I had some trouble reproducing the quantitative results on NTU-RGBD dataset with lie. I found that if I set `num_samples` to `num_motions`, which is 13 for this dataset, the FID would be about 20, way higher than the result in the paper. I also tried setting `replacement` to False, and got FID around 0.76.
> > Could you kindly provide a more detailed dataloader code for ground truth as well as the generated results? Thanks!
> 
> Hi, the codes of Dataloaders & Datasets could be found in this repo https://github.com/EricGuo5513/action2motion/tree/master/motion_loaders .Please note "Num Samples/Motions" means the number of motions we want to sample/generate. It should be 3000 for all datasets, and replacement is set to True in our experiments. "13" for NTU-RGBD only means the "Number of Action Categories".

Thanks for replying! Using the code you provided, I've reproduced the results on NTU-RGBD, except for the FID of ground truth. I only got about 0.06, while the paper says it's 0.031+-0.004. I wonder if there's any changes you made when calculating the ground truth FID? Here are the logs:
==================== Replication 2 ====================
Time: 2021-08-25 21:54:14.610624
========== Evaluating Accuracy ==========
Calculating Accuracies...
---> [ground truth] Accuracy: 0.9997
Calculating Accuracies...
---> [vanilla_vae_lie_mse_kld01] Accuracy: 0.9510
Time: 2021-08-25 21:54:19.936147
========== Evaluating FID ==========
Calculating Activations...
Calculating Activations...
=== Evaluating Diversity ===
=== Evaluating Multimodality ===
---> [ground truth] FID: 0.0577
---> [ground truth] Diversity: 7.3130
---> [ground truth] Multimodality: 2.1393
Calculating Activations...
=== Evaluating Diversity ===
=== Evaluating Multimodality ===
---> [vanilla_vae_lie_mse_kld01] FID: 0.3006
---> [vanilla_vae_lie_mse_kld01] Diversity: 6.9541
---> [vanilla_vae_lie_mse_kld01] Multimodality: 2.0697

**Comment by EricGuo5513** [comment]
Hi, I think 0.06 should be in a normal range. If you worry about the rigidity, you could report our result as well as your reproduced results as in [Action-Conditioned 3D Human Motion Synthesis with Transformer VAE](https://arxiv.org/pdf/2104.05670.pdf).

**Comment by sjxAndy** [comment]
> Hi, I think 0.06 should be in a normal range. If you worry about the rigidity, you could report our result as well as your reproduced results as in [Action-Conditioned 3D Human Motion Synthesis with Transformer VAE](https://arxiv.org/pdf/2104.05670.pdf).

I see. Thank you!
